bootstrap/src/basis/ordmap.ml

```diff
@@ -324,9 +324,137 @@ module T = struct
   end
 end
 include T
-include Container_common.Make_poly3_fold(T)
 include Container_array.Make_poly3_array(T)
 
+let fold_until ~init ~f t =
+  let rec fn accum f = function
+    | Empty -> accum, false
+    | Leaf {k; v} -> f accum (k, v)
+    | Node {l; k; v; n=_; h=_; r} -> begin
+        match fn accum f l with
+        | accum', true -> accum', true
+        | accum', false -> begin
+            match f accum' (k, v) with
+            | accum'', true -> accum'', true
+            | accum'', false -> fn accum'' f r
+          end
+      end
+  in
+  let accum, _ = fn init f t.root in
+  accum
+
+let fold_right_until ~init ~f t =
+  let rec fn accum f = function
+    | Empty -> accum, false
+    | Leaf {k; v} -> f (k, v) accum
+    | Node {l; k; v; n=_; h=_; r} -> begin
+        match fn accum f r with
+        | accum', true -> accum', true
+        | accum', false -> begin
+            match f (k, v) accum' with
+            | accum'', true -> accum'', true
+            | accum'', false -> fn accum'' f l
+          end
+      end
+  in
+  let accum, _ = fn init f t.root in
+  accum
+
+let foldi_until ~init ~f t =
+  let _, accum = fold_until t ~init:(0, init)
+    ~f:(fun (i, accum) (k, v) ->
+      let i' = (Usize.succ i) in
+      let accum', until = f i accum (k, v) in
+      (i', accum'), until
+    ) in
+  accum
+
+let fold ~init ~f t =
+  fold_until t ~init ~f:(fun accum (k, v) -> (f accum (k, v)), false)
+
+let fold_right ~init ~f t =
+  fold_right_until t ~init ~f:(fun (k, v) accum -> (f (k, v) accum), false)
+
+let foldi ~init ~f t =
+  foldi_until t ~init ~f:(fun i accum (k, v) -> (f i accum (k, v)), false)
+
+let iter ~f t =
+  fold t ~init:() ~f:(fun _ (k, v) -> f (k, v))
+
+let iteri ~f t =
+  foldi t ~init:() ~f:(fun i _ (k, v) -> f i (k, v))
+
+let count ~f t =
+  fold t ~init:0 ~f:(fun accum (k, v) ->
+    match f (k, v) with
+    | false -> accum
+    | true -> (Usize.succ accum)
+  )
+
+let for_any ~f t =
+  fold_until t ~init:false ~f:(fun _ (k, v) ->
+    let any' = f (k, v) in
+    any', any'
+  )
+
+let for_all ~f t =
+  fold_until t ~init:true ~f:(fun _ (k, v) ->
+    let all' = f (k, v) in
+    all', (not all')
+  )
+
+let find ~f t =
+  fold_until t ~init:None ~f:(fun _ (k, v) ->
+    match f (k, v) with
+    | false -> None, false
+    | true -> Some (k, v), true
+  )
+
+let find_map ~f t =
+  fold_until t ~init:None ~f:(fun _ (k, v) ->
+    match f (k, v) with
+    | None -> None, false
+    | Some a -> Some a, true
+  )
+
+let findi ~f t =
+  foldi_until t ~init:None ~f:(fun i _ (k, v) ->
+    match f i (k, v) with
+    | false -> None, false
+    | true -> Some (k, v), true
+  )
+
+let findi_map ~f t =
+  foldi_until t ~init:None ~f:(fun i _ (k, v) ->
+    match f i (k, v) with
+    | None -> None, false
+    | Some a -> Some a, true
+  )
+
+let min_elm ~cmp t =
+  fold t ~init:None ~f:(fun accum (k, v) ->
```

Comment:
I don't quite understand why we have to do a full fold here (same for `max_elm`). If this is an ordmap, can't you just take the leftmost elm here? Basically, a `fold_until` that returns false on the first elm?

Reply:
Looks like the docs are confusing.  The intent is to return the leftmost element in the set of elements which would return `Lt` or `Eq` with respect to all other elements.  In other words, a full pass is required, but if multiple minima exist, the leftmost one is returned rather than an arbitrary one.

Reply:
Oh, my mistake. I mixed this up with key comparison, no value comparison.